File: hardcoded/gadgets/mov.py
from gadgets import gadget, util
import ropchain


def find(r1, r2, gadgets, canUse):
    rop = gadget.find(gadgets, 'mov', r1, r2)
    if rop != None:
        rop = ropchain.ROPChain(rop)
        return rop
    rop = util.optMap(rop, fromLea, r1, r2, gadgets, canUse)
    rop = util.optMap(rop, fromLeaWithOffset, r1, r2, gadgets, canUse)
    rop = util.optMap(rop, fromXorXor, r1, r2, gadgets, canUse)
    rop = util.optMap(rop, fromXorOr, r1, r2, gadgets, canUse)
    rop = util.optMap(rop, fromXchg, r1, r2, gadgets, canUse)
    return rop


'''
lea r1, [r2]
'''
def fromLea(r1, r2, gadgets, canUse):
    _lea = lea.find(r1, '[%s]' % r2, gadgets, canUse)
    if _lea != None:
        return ropchain.ROPChain(_lea)
    else:
        return None

'''
lea r1, [r2+n]
(dec r1) * n
'''
def fromLeaWithOffset(r1, r2, gadgets, canUse):
    #TODO replace find.ByRegex with wrapper function of lea
    _lea, _, imm = gadget.findByRegex(gadgets, 'lea', r1, r'\[%s\+0x[0-9a-fA-F]*]')
    _dec = dec.find(r1, gadgets, canUse)
    if _lea != None and _dec != None:
        return ropchain.ROPChain([_lea] +  [_dec] * imm)
    else:
        return None

'''
xor r1, r1;
xor r1, r2;
'''
def fromXorXor(r1, r2, gadgets, canUse):
    xorR1R1 = xor.find(r1, r1, gadgets, canUse)
    xorR1R2 = xor.find(r1, r2, gadgets, canUse)
    if xorR1R1 != None and xorR1R2 != None:
        return ropchain.ROPChain([xorR1R1, xorR1R2])
    else:
        return None

'''
xor r1, r1
or r1, r2
'''
def fromXorOr(r1, r2, gadgets, canUse):
    xorR1R1 = xor.find(r1, r1, gadgets, canUse)
    orGadget = orGadget.find(r1, r2, gadgets, canUse)
    if xorR1R1 != None and orGadget != None:
        return ropchain.ROPChain([xorR1R1, orGadget])
    else:
        return None

'''
xchg r1, r2
mov r2, r1
xchg r1, r2
'''
def fromXchg(r1, r2, gadgets, canUse):
    _xchg = xchg.find(r1, r2, gadgets, canUse)
    _mov = mov.find(r2, r1, gadgets, canUse)
    if _xchg != None and _mov != None:
        return ropchain.ROPChain([_xchg, _mov, _xchg])
    else:
        return None

File: ropchain/lib/src/asm/mov.cpp
#include "mov.h"
#include "xor.h"
#include "or.h"
#include "xchg.h"

OptROP fromLea(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval);
OptROP fromLeaWithOffset(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval);
OptROP fromXorXor(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval);
OptROP fromXorOr(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval);
OptROP fromXchg(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval);

OptROP findWithoutXchg(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval) {
    auto gadget = Util::find(gadgets, aval, "mov", op1, op2);
    auto rop = Util::toOptROP(gadget);
    rop = Util::optMin(rop, fromLea(op1, op2, gadgets, aval));
    rop = Util::optMin(rop, fromLeaWithOffset(op1, op2, gadgets, aval));
    rop = Util::optMin(rop, fromXorXor(op1, op2, gadgets, aval));
    rop = Util::optMin(rop, fromXorOr(op1, op2, gadgets, aval));
    return rop;
}

OptROP Mov::find(const Opcode& op1, const Opcode& op2,
        const Gadgets& gadgets, RegSet& aval) {
    auto rop = findWithoutXchg(op1, op2, gadgets, aval);
    rop = Util::optMin(rop, fromXchg(op1, op2, gadgets, aval));
    return rop;
}

//lea r1, [r2]; ret
OptROP fromLea(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval) {
    //TODO
    return {};
}

//xor r1, r1; ret; or r1, r2; ret
OptROP fromXorOr(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval) {
    auto xorR1R2 = Xor::find(op1, op2, gadgets, aval);
    auto orR1R2 = Or::find(op1, op2, gadgets, aval);
    if(xorR1R2.has_value() && orR1R2.has_value()) {
        return xorR1R2.value() + orR1R2.value();
    }
    return {};
}

//lea r1, [r2+imm]; ret; (dec r1; ret)*
OptROP fromLeaWithOffset(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval) {
    //TODO
    return {};
}

//xor r1, r1; ret; xor r1, r2; ret
OptROP fromXorXor(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval) {
    auto xorR1R1 = Xor::find(op1, op1, gadgets, aval);
    auto xorR1R2 = Xor::find(op1, op2, gadgets, aval);
    if(xorR1R1.has_value() && xorR1R2.has_value()) {
        return xorR1R1.value() + xorR1R2.value();
    }
    return {};
}

//xchg r1, r2; ret mov r2, r1; ret; xchg r1, r2; ret
OptROP fromXchg(const Opcode& op1, const Opcode& op2, const Gadgets& gadgets, RegSet& aval) {
    auto xchg = Xchg::find(op1, op2, gadgets, aval);
    auto mov = findWithoutXchg(op2, op1, gadgets, aval);
    if(xchg.has_value() && mov.has_value()) {
        return xchg.value() + mov.value() + xchg().value();
    }
}
